### DISCUSSION TITLE
Typegen doesn't work for entry/exit actions

### DISCUSSION THREAD

**Discussion by Ghirigoro** [statelyai/xstate#3776] [Q&A] [answered]
I'm passing the initial state to a machine that uses typegen and it works fine except for actions. For example:
```ts
function createTestMachine(state: State) {
    return createMachine(
        {
            tsTypes: {} as import("./test.fsm.typegen").Typegen0,
            schema: {
                context: {} as Context,
                events: {} as TestEvent,
            },
            id: "admin",
            initial: state.current, // This is a string that corresponds to the state names
            states: {
                a: {
                    entry: ["failingAction"],
                },
                b: {
                    on: {
                        NEXT: {
                            target: "a",
                            actions: ["workingAction"],
                        },
                    },
                },
            },
        },
        {
            actions: {
                failingAction: () => {},// ERROR: TS complains here.
                workingAction: () => {},
            },
        }
    )
}
```
If I give the machine a static initial value then things start to work again. Is this a known limitation? A bug? Anyone have workarounds?

**Comment by Andarist** [comment, marked as answer]
This is a known limitation. The problem with that approach is that currently, we can't even visualize it correctly - where we'd draw the "initial state marker"?

We gonna explore improved support for patterns like this in the future but it won't happen any time soon.

**Reply by Ghirigoro** [reply]
Thanks for the reply. Not sure it's on the roadmap, but it would be nice to have some mechanism to create a machine in a particular state (resuming an ongoing interaction is a pretty common use case). What I'm doing currently is creating an initial `init` state that routes the machine to a state based on context but it's pretty verbose and error prone (i.e. every time I change states I have to make sure to update that state):

```ts
init: {
  always: [
    { cond: "isInStateX", target: "X" },
    { cond: "isInStateY", target: "Y" },
    // etc...
  ]
}
```
There's probably a better way but I haven't found it yet ;)

**Reply by Andarist** [reply]
I believe that in v5 (the upcoming major version) we already have support for this:
```js
init: {
  initial: [
    { cond: "isInStateX", target: "X" },
    { cond: "isInStateY", target: "Y" },
    // etc...
  ]
}
```

So it doesn't have to rely on `always` transitions. `always` is a good approximation of that feature but keep in mind that `always` might be checked after every event received in the given state - `initial` would be checked only when entering that state.

I totally understand though that sometimes it's just not super feasible to list all possible states within such `always`/`initial` conditions. I think there is some potential in allowing more "dynamic" targets - but I'm not yet sure what that would look like and if it's even a good idea. We might explore this in the future to learn more about its pros and cons.
